Use only string values from the cookiecutter context

oncleben31 opened the pull request:
fix: #166

oncleben31 committed:
Add test for filter method

Check all values that are not strings are ignored in the cookiecutter context.

oncleben31 committed:
Fix to ignore not strings values

oncleben31 committed:
Fix type hint errors from mypy

cjolowicz commented on src/retrocookie/filter.py:
```diff
@@ -33,6 +34,7 @@ def ref(key: str) -> str:
         for key, value in context.items()
         if key not in exclude_variables
+        and isinstance(value, str)
         and not (include_variables and key not in include_variables)
```
Can we reorder this so the checks for `include_variables` and `exclude_variables` are next to each other?

```suggestion
        and not (include_variables and key not in include_variables)
        and isinstance(value, str)
```

cjolowicz commented on src/retrocookie/core.py:
```diff
@@ -23,11 +25,13 @@ def find_template_directory(repository: git.Repository) -> Path:
     path = Path(".cookiecutter.json")
     text = repository.read_text(path, ref=ref)
-    return cast(Dict[str, str], json.loads(text))
+    return cast(Dict[str, Union[str, List[Any], None]], json.loads(text))
```
Actually, the best we can do here is `Dict[str, Any]`, and we need to perform runtime type checking to enforce this before the cast:

```suggestion
    data = json.loads(text)
    if not isinstance(data, dict) or not all(isinstance(key, str) for key in data):
        raise TypeError(".cookiecutter.json does not contain an object")
    return cast(Dict[str, Any], data)
```

To elaborate on this: The return value of `json.loads` can be any valid Python representation of JSON. As there is currently no recursive type support in mypy that would allow us to express JSON types, we can't do much better than `Any`. However, we know that `.cookiecutter.json` *should* contain a JSON object, i.e. a mapping of strings to JSON values. So we need to check that the input conforms to this expectation, and return `Dict[str, Any]` if it does.

The type signatures of this and the other functions would need to change accordingly.

cjolowicz reviewed (changes requested):
Many thanks! :+1: :taco:   Please see the inline comments.

oncleben31 committed:
Reorder conditions

Co-authored-by: Claudio Jolowicz <cjolowicz@gmail.com>

oncleben31 committed:
Check data consistency in .cookiecutter.json

Co-authored-by: Claudio Jolowicz <cjolowicz@gmail.com>

oncleben31 committed:
Fix type hint

oncleben31 commented:
I need to add tests for having 100% code coverage, but it's difficult to do it locally as I don't succeed to run tests session with nox. Most of the tests fail with a `_pygit2.GitError: reference 'refs/heads/main' not found` error

I suspect an error due to git version or configuration (as it's the only thing not pinned by poetry/nox). But I have no clue to find a solution.

oncleben31 commented:
I was right the command `git config --global init.defaultbranch master` allows to run the nox session without failure.
I'm using `main` as default. Do I need to track this error in this repository or should it be open in pygit2 repository ?

**Update**: After second though and some look at the code, I think most of the code use `master` as default if no specific branch is given. So I think the issue is in the test suite.

oncleben31 committed:
Add test for load_context error case

load_context raises a TypeError when .cookiecutter.json is not a JSON
dictionnary.

oncleben31 committed:
Fix code style and mypy error
